src/traits/networking/web_sever_libp2p_fallback.rs

```diff
@@ -0,0 +1,254 @@
+                num_bootstrap,
+                network_id,
+            ),
+            <Libp2pNetwork<Message<TYPES, I>, TYPES::SignatureKey> as TestableNetworkingImplementation<_, _>>::generator(expected_node_count, num_bootstrap, network_id)
+        );
+        Box::new(move |node_id| {
+            CombinedNetworks(
+                generators.0(node_id),
+                generators.1(node_id),
+                PhantomData::default(),
+            )
+        })
+    }
+
+    /// Get the number of messages in-flight.
+    ///
+    /// Some implementations will not be able to tell how many messages there are in-flight. These implementations should return `None`.
+    fn in_flight_message_count(&self) -> Option<usize> {
+        None
+    }
+}
+
+#[async_trait]
+impl<
+        TYPES: NodeType,
+        I: NodeImplementation<TYPES>,
+        PROPOSAL: ProposalType<NodeType = TYPES>,
+        VOTE: VoteType<TYPES>,
+        MEMBERSHIP: Membership<TYPES>,
+    > CommunicationChannel<TYPES, Message<TYPES, I>, PROPOSAL, VOTE, MEMBERSHIP>
+    for WebServerWithFallbackCommChannel<TYPES, I, PROPOSAL, VOTE, MEMBERSHIP>
+{
+    type NETWORK = CombinedNetworks<TYPES, I, PROPOSAL, VOTE, MEMBERSHIP>;
+
+    async fn wait_for_ready(&self) {
+        join!(
+            self.network().wait_for_ready(),
+            self.fallback().wait_for_ready()
+        );
+    }
+
+    async fn is_ready(&self) -> bool {
+        self.network().is_ready().await && self.fallback().is_ready().await
+    }
+
+    async fn shut_down(&self) -> () {
+        join!(self.network().shut_down(), self.fallback().shut_down());
+    }
+
+    async fn broadcast_message(
+        &self,
+        message: Message<TYPES, I>,
+        election: &MEMBERSHIP,
+    ) -> Result<(), NetworkError> {
+        let recipients =
+            <MEMBERSHIP as Membership<TYPES>>::get_committee(election, message.get_view_number());
+        let fallback = self
+            .fallback()
+            .broadcast_message(message.clone(), recipients.clone());
+        let network = self.network().broadcast_message(message, recipients);
+        match join!(fallback, network) {
+            (Err(e), Err(_)) => Err(e),
+            _ => Ok(()),
+        }
+    }
+
+    async fn direct_message(
+        &self,
+        message: Message<TYPES, I>,
+        recipient: TYPES::SignatureKey,
+    ) -> Result<(), NetworkError> {
+        match self
+            .network()
+            .direct_message(message.clone(), recipient.clone())
+            .await
+        {
+            Ok(_) => Ok(()),
+            Err(_e) => self.fallback().direct_message(message, recipient).await,
+        }
+    }
+
+    async fn recv_msgs(
+        &self,
+        transmit_type: TransmitType,
+    ) -> Result<Vec<Message<TYPES, I>>, NetworkError> {
+        match self.network().recv_msgs(transmit_type).await {
+            Ok(msgs) => Ok(msgs),
+            Err(_e) => self.fallback().recv_msgs(transmit_type).await,
```

Comment:
Probably something to push to a future sprint, but we may want to try listening on both channels with a timeout in case the centralized server is down for other nodes.

Reply:
I've been thinking a bunch on this one, I think we should probably should be clearing both queues and on each call here.  Otherwise the backup queue will just keep growing, and we won't really know if we missed something.  

I'm actually going to tackle this one this sprint, probably next.  This needs something a bit more sophisticated